### PR TITLE
WIP: Refactor all the git

### PR DESCRIPTION
**Pull request by zeripath**
This PR is a working branch for refactoring the entirety of modules/git to allow for a decision between using a native git backend and a go-git backend at runtime rather than build time.

THIS IS VERY MUCH WIP AND IS INCOMPLETE. THERE WILL BE FORCE PUSHES AT POINT AND IS NOT EXPECTED TO WORK AT PRESENT.

### CODE REVIEW

**Comment by vtolstov** [comment]
Very cool!

**Comment by 6543** [comment]
I think this will need a rebase ;)

File: models/pull_sign.go [hunk 1]
```diff
@@ -6,12 +6,14 @@ package models
 
 import (
 	"code.gitea.io/gitea/modules/git"
+	"code.gitea.io/gitea/modules/git/service"
+	gitservice "code.gitea.io/gitea/modules/git/service"
```
Comment (6543) [inline models/pull_sign.go:10]: same import with diff name

File: models/pull_sign.go [hunk 2]
```diff
@@ -6,12 +6,14 @@ package models
...
 	"code.gitea.io/gitea/modules/log"
 	"code.gitea.io/gitea/modules/setting"
 )
 
 // SignMerge determines if we should sign a PR merge commit to the base repository
-func (pr *PullRequest) SignMerge(u *User, tmpBasePath, baseCommit, headCommit string) (bool, string, *git.Signature, error) {
+func (pr *PullRequest) SignMerge(u *User, tmpBasePath, baseCommit, headCommit string) (bool, string, *service.Signature, error) {
```
Comment (6543) [inline models/pull_sign.go:16]:
```suggestion
func (pr *PullRequest) SignMerge(u *User, tmpBasePath, baseCommit, headCommit string) (bool, string, *gitservice.Signature, error) {
```

File: models/pull_sign.go [hunk 3]
```diff
@@ -6,12 +6,14 @@ package models
 
 import (
 	"code.gitea.io/gitea/modules/git"
+	"code.gitea.io/gitea/modules/git/service"
```
Comment (6543) [inline models/pull_sign.go:9]:
```suggestion
```

File: services/archiver/archiver.go [hunk 4]
```diff
@@ -18,6 +18,8 @@ import (
 	"code.gitea.io/gitea/modules/base"
 	"code.gitea.io/gitea/modules/context"
 	"code.gitea.io/gitea/modules/git"
+	"code.gitea.io/gitea/modules/git/service"
```
Comment (6543) [inline services/archiver/archiver.go:21]:
```suggestion
```

File: services/archiver/archiver.go [hunk 5]
```diff
@@ -31,13 +33,13 @@ import (
 // handle elsewhere.
 type ArchiveRequest struct {
 	uri             string
-	repo            *git.Repository
+	repo            service.Repository
```
Comment (6543) [inline services/archiver/archiver.go:36]:
```suggestion
	repo            gitservice.Repository
```

File: services/archiver/archiver.go [hunk 6]
```diff
@@ -31,13 +33,13 @@ import (
...
 	refName         string
 	ext             string
 	archivePath     string
-	archiveType     git.ArchiveType
+	archiveType     gitservice.ArchiveType
 	archiveComplete bool
-	commit          *git.Commit
+	commit          service.Commit
```
Comment (6543) [inline services/archiver/archiver.go:42]:
```suggestion
	commit          gitservice.Commit
```

File: services/archiver/archiver.go [hunk 7]
```diff
@@ -99,10 +101,10 @@ func (aReq *ArchiveRequest) TimedWaitForCompletion(ctx *context.Context, dur tim
 }
 
 // The caller must hold the archiveMutex across calls to getArchiveRequest.
-func getArchiveRequest(repo *git.Repository, commit *git.Commit, archiveType git.ArchiveType) *ArchiveRequest {
+func getArchiveRequest(repo service.Repository, commit service.Commit, archiveType gitservice.ArchiveType) *ArchiveRequest {
```
Comment (6543) [inline services/archiver/archiver.go:104]:
```suggestion
func getArchiveRequest(repo gitservice.Repository, commit gitservice.Commit, archiveType gitservice.ArchiveType) *ArchiveRequest {
```

**Comment by delvh** [comment]
Is this PR still maintained?
I don't think so.
Feel free to reopen if you're still wanting to work on it.

**Pull request closed without merging**